game-core/src/main/java/games/strategy/triplea/attachments/CanalAttachment.java

```diff
@@ -31,24 +32,29 @@ public CanalAttachment(final String name, final Attachable attachable, final Gam
     super(name, attachable, gameData);
   }
 
-  public static Set<Territory> getAllCanalSeaZones(final String canalName, final GameData data) {
-    final Set<Territory> territories = new HashSet<>();
-    for (final Territory t : data.getMap()) {
-      final Set<CanalAttachment> canalAttachments = get(t);
-      if (canalAttachments.isEmpty()) {
-        continue;
-      }
-      for (final CanalAttachment canalAttachment : canalAttachments) {
+  /**
+   * Checks if the route contains both territories to pass through the given canal. If route is
+   * null returns true.
+   */
+  public static boolean isCanalOnRoute(final String canalName, final Route route) {
+    if (route == null) {
+      return true;
+    }
+    boolean previousTerritoryHasCanal = false;
+    for (final Territory t : route) {
+      boolean currentTerritoryHasCanal = false;
+      for (final CanalAttachment canalAttachment : get(t)) {
         if (canalAttachment.getCanalName().equals(canalName)) {
-          territories.add(t);
+          currentTerritoryHasCanal = true;
+          break;
         }
       }
+      if (previousTerritoryHasCanal && currentTerritoryHasCanal) {
+        return true;
+      }
+      previousTerritoryHasCanal = currentTerritoryHasCanal;
     }
-    if (territories.size() != 2) {
-      throw new IllegalStateException(
-          "Wrong number of sea zones for canal (exactly 2 sea zones may have the same canalName):" + territories);
```

Comment:
This validation was moved down to the parsing validate method as it should be checked during map parsing not every time a route is calculated.